chore(dockerfile): make docker build multistage

hazzadous opened the pull request:
The built image is >4GB uncompressed atm, I'm pretty sure there is a lot
of cruft.

Plan is to split out the django, frontend, plugin-server builds and
hopefully get some gains in there to not include build deps.

This gets the uncompressed image down to 1.48GB, compressed around 500MB 🤞 

It also adds caching to the test image build workflow.

Some important changes there are slightly tangential:

 1. we run on master branch, to ensure we have the master image for caching
 2. we run the E2E tests in the same workflow, as using `workflow_run` trigger events seemed to be more complicated than expected
 3. to ensure that the image is functional, we run the E2E tests with it

NOTE: I've also kicked off a test with the helm chart [here](https://github.com/PostHog/charts-clickhouse/pull/449/checks)
NOTE 2: if caching works, the total runtime of both the build and E2E is ~10mins. If it doesn't though, it's 30 mins. E2E previously I think is around 20 mins currently.

TODO:

 * distribute staticfiles only by CDN(? this might not work for super security conscious installs)
 * distribute a separate plugin server image
 * we are doubling up on staticfiles with the collect step, we can probably get rid of the originals somehow but I'm not sure exactly how
 * clean up the user handling, we switch back an forth but could probably do better
 * use this image for tests, to ensure the image build is correct
 * caching is good sometimes, but not if you want security patches to be applied. Make sure prod has security patches applied frequently. I've actually only enabled caching on branches so that's not an issue atm.
 * remove old image tags so orphan layers can be pruned
 * We also build the image elsewhere for releases, we can just reuse this build instead

## Problem

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

Commit:
chore(dockerfile): make docker build multistage

The built image is >4GB uncompressed atm, I'm pretty sure there is a lot
of cruft.

Plan is to split out the django, frontend, plugin-server builds and
hopefully get some gains in there to not include build deps.

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
fix dockerfile lints

Commit:
cache from branch

pauldambra commented on .github/workflows/docker-image-test.yml:
```diff
@@ -16,19 +16,57 @@ jobs:
-                  push: false
-                  load: true
-                  tags: posthog/posthog:testing
+                  push: true
```
`load:true` was needed so the output image size job worked

_but_ at a quarter of the size maybe we just delete the output image size job

hazzadous replied:
ahhhh ok I'll revert

Commit:
add load :true

fuziontech reviewed (approved):
This looks really exciting to me. Let's land this on Monday and get this in! 🚢 it

Commit:
Update production.Dockerfile

Co-authored-by: James Greenhill <fuziontech@gmail.com>

hazzadous commented on production.Dockerfile:
```diff
@@ -15,25 +65,15 @@ WORKDIR /code
-    "libxml2-dev~=2.9" \
-    "xmlsec~=1.2" \
-    "xmlsec-dev~=1.2"
+# Install SAML runtime dependencies
+RUN apk --update --no-cache add "xmlsec~=1.2"
```
```suggestion
```

Commit:
Update production.Dockerfile

Commit:
update to use compressed size from remote repo

Commit:
tag with branch and sha

Commit:
add ref on pull_request events

Commit:
install python

Commit:
be a bit more lax with python version

Commit:
fix image size calc

Commit:
hardcode lower case image name

Commit:
use @

Commit:
only add sha on master branch, add master tag on master

guidoiaquinti commented on production.Dockerfile:
```diff
@@ -3,6 +3,57 @@
+
+WORKDIR /code/plugin-server
+
+# Install python, make and gcc as they are needed for the yarn install
```
Nit: also Python?

hazzadous replied:
As far as I could tell yes python is also needed. I could be wrong there though.

guidoiaquinti commented:
Still quite a mess but way better than before. Nice work (as usual) @hazzadous !

Commit:
chore: use docker image in e2e tests

This is to try to add some guarantees to the docker image without having
to manually test it, so we can be a bit more aggressive with
improvements without, e.g. having to push to the playground or run
locally.

Commit:
wip

Commit:
add to summary

Commit:
wip

Commit:
chore: put cypress tests after docker build

I couldn't figure out a way to get workflow_run to run without merging
in, so I'm just putting after the build.

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
remove quotes

Commit:
remove separate cypress install

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
add gunicorn.config.py

Commit:
ci: run docker image build on master as well

This way we get the caching from the master build.

Commit:
wip

Commit:
wip

benjackwhite commented on production.Dockerfile:
```diff
@@ -46,70 +85,73 @@ RUN apk --update --no-cache add \
-    chown -R posthog:1000 /home/posthog/code
-WORKDIR /home/posthog/code
+# Add in the compiled plugin-server
+COPY --from=plugin-server /code/plugin-server/dist/ ./dist/
```
Is it really the case that the plugin-server does not need `node_modules` ? I understand it for the frontend as that is all bundled code but surely nodejs uses it in the plugin server?

hazzadous replied:
I suspect it does need them, so we install them in the final image as well.

Note that we now also use the image in the E2E tests, which should catch this(?)

hazzadous replied:
( see https://github.com/PostHog/posthog/pull/10488/files#diff-55d06c59f3552514747959fc163ad2940d33f0757fa48cf363580af437dee6a3R131 )
